main.go

```diff
@@ -23,12 +23,16 @@ func main() {
 	c := prepareConfig()
 	log.Printf("configuration: %+v", c)
 
-	client, err := connectMqtt(c.mqtt.server)
+	client, err := connectMqtt(c)
 
 	if err != nil {
 		panic(err)
 	}
 
+	monitorSensor(c, client)
```

Comment:
Few things here, 
1. Instead of passing through the entire config structure, it would be to pass the individual values through as explicit args, or even better, purely the `configSensor` struct value.

```go
monitorSensor(c.sensor, client)
```

This will mean you could call monitorSensor different times with
```go
monitorSensor(configSensor{pin:"5", checkRate: time.Second, topic: "blah"}, client)
```

2. monitorSensor now contains a blocking loop (infinite) that stops the program "falling through". This is a bit of a no-no because effectively you can only ever call monitorSensor once like you have here. 
The way around this is to call monitorSensor as a **go routine** however this will mean your blocking loop is now contained within the go routine and thus your program will exit a few MS after you start it.

To get around this problem we want to block the application from closing until it receives an os signal

```go 
sigChan := make(chan os.Signal, 1) // make a channel that receives the os Signal it receives
signal.Notify(sigChan, os.Kill) // when the application receives any signal it deems as kill, it will write this into the channel

// your logic
go monitorSensor(c.sensor, client)

sig := <-sigChan // sit here and wait for data in the channel. When the channel eventually receives data, we will continue from this line
// now you can write any logic you want here to tidy the application before it closes
client.Close()
pin.DoSomethingHere()

```

3. If you do the two previous steps, this means we can have an array of sensors to poll by updating the config to allow an array of sensors +

```go
sigChan := make(chan os.Signal, 1) // make a channel that receives the os Signal it receives
signal.Notify(sigChan, os.Kill) // when the application receives any signal it deems as kill, it will write this into the channel

// your logic
for _, sensor := range c.sensors {
  go monitorSensor(sensor, client)
}

sig := <-sigChan
log.PrintF("application terminated")
```